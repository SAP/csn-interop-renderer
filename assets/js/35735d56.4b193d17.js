"use strict";(self.webpackChunk_sap_csn_interop_renderer_website=self.webpackChunk_sap_csn_interop_renderer_website||[]).push([[775],{2082:e=>{e.exports=JSON.parse('{"version":{"pluginId":"default","version":"current","label":"Next","banner":null,"badge":false,"noIndex":false,"className":"docs-version-current","isLast":true,"docsSidebars":{"defaultSidebar":[{"type":"link","label":"Overview","href":"/csn-interop-renderer/","docId":"index","unlisted":false},{"type":"category","label":"Integrations","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"NPM package","href":"/csn-interop-renderer/Integrations/npm","docId":"Integrations/npm","unlisted":false},{"type":"link","label":"Web Component","href":"/csn-interop-renderer/Integrations/web-component","docId":"Integrations/web-component","unlisted":false},{"type":"link","label":"Using CLI","href":"/csn-interop-renderer/Integrations/cli","docId":"Integrations/cli","unlisted":false}]},{"type":"category","label":"examples","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Airline (Example)","href":"/csn-interop-renderer/examples/Airline","docId":"examples/Airline","unlisted":false},{"type":"link","label":"BillingDocument","href":"/csn-interop-renderer/examples/BillingDocument","docId":"examples/BillingDocument","unlisted":false},{"type":"link","label":"CostCenter","href":"/csn-interop-renderer/examples/CostCenter","docId":"examples/CostCenter","unlisted":false},{"type":"link","label":"CostCenterCombined","href":"/csn-interop-renderer/examples/CostCenterCombined","docId":"examples/CostCenterCombined","unlisted":false},{"type":"link","label":"GeneralLedgerAccountLineItem","href":"/csn-interop-renderer/examples/GeneralLedgerAccountLineItem","docId":"examples/GeneralLedgerAccountLineItem","unlisted":false}]}]},"docs":{"examples/Airline":{"id":"examples/Airline","title":"Airline (Example)","description":"This CSN example document shows how the airline example is expressed with a CDS Service exposing the entities through an API.","sidebar":"defaultSidebar"},"examples/BillingDocument":{"id":"examples/BillingDocument","title":"BillingDocument","description":"Entity Definitions","sidebar":"defaultSidebar"},"examples/CostCenter":{"id":"examples/CostCenter","title":"CostCenter","description":"Entity Definitions","sidebar":"defaultSidebar"},"examples/CostCenterCombined":{"id":"examples/CostCenterCombined","title":"CostCenterCombined","description":"Entity Definitions","sidebar":"defaultSidebar"},"examples/GeneralLedgerAccountLineItem":{"id":"examples/GeneralLedgerAccountLineItem","title":"GeneralLedgerAccountLineItem","description":"Entity Definitions","sidebar":"defaultSidebar"},"index":{"id":"index","title":"Overview","description":"Motivation","sidebar":"defaultSidebar"},"Integrations/cli":{"id":"Integrations/cli","title":"Using CLI","description":"Suitable for local testing on your machine.","sidebar":"defaultSidebar"},"Integrations/npm":{"id":"Integrations/npm","title":"NPM package","description":"Suitable for JavaScript/TypeScript based projects.","sidebar":"defaultSidebar"},"Integrations/web-component":{"id":"Integrations/web-component","title":"Web Component","description":"Suitable for Frontend e.g. React/Angular based projects.","sidebar":"defaultSidebar"}}}}')}}]);