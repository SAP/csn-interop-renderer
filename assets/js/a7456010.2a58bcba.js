"use strict";(self.webpackChunk_sap_csn_interop_renderer_website=self.webpackChunk_sap_csn_interop_renderer_website||[]).push([[235],{8552:e=>{e.exports=JSON.parse('{"name":"docusaurus-plugin-content-pages","id":"default"}')}}]);