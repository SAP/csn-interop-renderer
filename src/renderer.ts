import {
  type TypeDefinition,
  type AssociationType,
  type CSNInteropRoot,
  type ElementEntry,
  type DefinitionEntry,
  type EntityDefinition,
  type StringTypeDefinition,
  type LargeStringTypeDefinition,
  type StructuredElementReference,
  type EqualsOperator,
  type ANDOperator,
  type OnValue,
  type CustomType,
  type ServiceDefinition,
} from "@sap/csn-interop-specification";
import { marked } from "marked";
import { gfmHeadingId } from "marked-gfm-heading-id";

const csnInteropEffectiveV1ExtensionsDocumentationPage =
  "https://sap.github.io/csn-interop-specification/spec-v1/extensions/";

// Helper functions
function isEntityDefinition(entry: DefinitionEntry): entry is EntityDefinition {
  return entry.kind === "entity";
}

function isStructuredElementReference(
  entry: StructuredElementReference | EqualsOperator | ANDOperator | OnValue,
): entry is StructuredElementReference {
  return typeof entry === "object" && Object.hasOwn(entry, "ref");
}

function isLengthConstrainable(
  entry: TypeDefinition | ElementEntry,
): entry is StringTypeDefinition | LargeStringTypeDefinition {
  return Object.hasOwn(entry, "length");
}

function isTypeDefinition(entry: DefinitionEntry): entry is TypeDefinition {
  return entry.kind === "type";
}

function isServiceDefinition(entry: DefinitionEntry): entry is ServiceDefinition {
  return entry.kind === "service";
}

function isCustomType(element: ElementEntry): element is CustomType {
  return !element.type.startsWith("cds.");
}

function isAssociationType(element: AssociationType | CustomType): element is AssociationType {
  return "target" in element && element.type === "cds.Association";
}

function renderContentWithI18n(content: unknown, i18n: CSNInteropRoot["i18n"] | undefined): string {
  const escapedContent = JSON.stringify(content).replace(/{/g, "&lbrace;").replace(/}/g, "&rbrace;");

  if (!i18n) {
    return escapedContent;
  }
  const matches = escapedContent.match(/^{i18n>(.*)}$/);
  const enKey = Object.keys(i18n).find((key) => key.match(/^en$/i));
  if (matches?.[1] && enKey && i18n[enKey]?.[matches[1]]) {
    return i18n[enKey][matches[1]];
  }
  return escapedContent;
}

/**
 * Get the header id the same format as GitHub does it for rendered markdown content.
 * @param definitionName The name of a definition
 * @param entityElementName The name of a definition element
 * @returns The markdown headline id in specific GitHub format style
 *
 */
function getHeaderId(definitionName: string, entityElementName?: string): string {
  if (entityElementName) {
    return `${definitionName.toLowerCase().replaceAll(".", "")}-${entityElementName.toLowerCase().replaceAll(".", "")}`;
  }
  return definitionName.toLowerCase().replaceAll(".", "");
}

function renderObject(content: object | null): string {
  return `<code>${JSON.stringify(content).replace(/{/g, "&lbrace;").replace(/}/g, "&rbrace;")}</code>`;
}

function getExtensionsDocumentationLink(_version?: string): string {
  return csnInteropEffectiveV1ExtensionsDocumentationPage;
}

// Sub render functions
function processEntities(
  entities: [string, EntityDefinition][],
  serviceNames: string[],
  i18n: CSNInteropRoot["i18n"] | undefined,
): string {
  if (!entities.length) return "";
  let output = "## Entity Definitions\n\n";
  for (const [entityName, entityDefinition] of entities) {
    output += `### ${entityName}\n\n`;
    if (entityDefinition.doc) {
      output += `${marked.parse(entityDefinition.doc)}\n`;
    }

    const dotSplittedEntityName = entityName.split(".");

    if (dotSplittedEntityName.length > 1 && serviceNames.includes(dotSplittedEntityName[0])) {
      output += `Exposed via:\n[${dotSplittedEntityName[0]}](#${dotSplittedEntityName[0].toLowerCase()})\n\n`;
    }

    output += `<table>\n`;

    output += `<tr>`;
    output += `<th>Element</th>`;
    output += `<th>Type</th>`;
    output += `<th>Description</th>`;
    output += `</tr>`;

    for (const [elementName, elementDefinition] of Object.entries(entityDefinition.elements)) {
      const annotations: [string, unknown][] = Object.entries(elementDefinition).filter(([key]) => key.startsWith("@"));
      const restProps = Object.entries(elementDefinition).filter(
        ([key]) => !key.startsWith("@") && !["doc", "type", "length", "cardinality", "on", "target"].includes(key),
      );

      const lengthConstraint =
        isLengthConstrainable(elementDefinition) && elementName.length ? `(${elementName.length})` : "";
      const typeLink = isCustomType(elementDefinition)
        ? `<a href="#${elementDefinition.type.toLowerCase()}">${elementDefinition.type}</a>`
        : elementDefinition.type + lengthConstraint;

      let description = "";

      const descriptionParts = restProps.map(
        ([key, value]) => `${key}: ${typeof value === "object" ? renderObject(value) : value}`,
      );

      // TODO: Composition handling

      // Association handling
      if (isAssociationType(elementDefinition) && elementDefinition.target) {
        const targetEntityName = elementDefinition.target;
        const onClause = elementDefinition.on;
        let targetElementName = "";
        let viaKey = "";

        // Extract foreign key and path descriptions
        if (Array.isArray(onClause)) {
          if (isStructuredElementReference(onClause[0])) {
            if (onClause[0].ref.length === 2) {
              targetElementName = onClause[0].ref[1];
            } else if (onClause[0].ref.length === 1) {
              viaKey = onClause[0].ref[0];
            }
          }
          if (isStructuredElementReference(onClause[2])) {
            if (onClause[2].ref.length === 2) {
              targetElementName = onClause[2].ref[1];
            } else if (onClause[2].ref.length === 1) {
              viaKey = onClause[2].ref[0];
            }
          }
        }

        const cardinality = elementDefinition.cardinality?.max === "*" ? "to many" : "to one";
        const pathDescription =
          targetEntityName && targetElementName
            ? `path: <a href="#${getHeaderId(targetEntityName)}">${targetEntityName}</a>.<a href="#${getHeaderId(targetEntityName, targetElementName)}">${targetElementName}</a>`
            : "";

        descriptionParts.push(
          `Association ${cardinality} <a href="#${getHeaderId(targetEntityName)}">${targetEntityName}</a> (${pathDescription}) via <a href="#${getHeaderId(entityName, viaKey)}">${viaKey}</a>`,
        );
      }

      const annotationParts = annotations.map(
        ([key, value]) =>
          `${key}: ${typeof value === "object" ? renderObject(value) : renderContentWithI18n(value, i18n)}`,
      );

      if (descriptionParts.length) {
        description += `${descriptionParts.join("<br />")}<br /><br />`;
      }
      if (annotationParts.length) {
        description += `<a href="${getExtensionsDocumentationLink()}" target="_blank">Annotations</a>:<br />`;
        description += `${annotationParts.join("<br />")}`;
      }

      output += `<tr>`;
      output += `<td><strong id="${getHeaderId(entityName, elementName)}">${elementName}</strong><br /><br /><small>${elementDefinition.doc ? elementDefinition.doc : ""}</small></td>`;
      output += `<td>${typeLink}</td>`;
      output += `<td>${description}</td>`;
      output += `</tr>\n`;
    }

    output += "</table>\n\n";
  }
  return output;
}

function processTypes(types: [string, TypeDefinition][], i18n: CSNInteropRoot["i18n"] | undefined): string {
  if (!types.length) return "";
  let output = "## Type Definitions\n\n";
  for (const [typeName, typeDefinition] of types) {
    output += `### ${typeName}\n\n`;
    if (typeDefinition.doc) {
      output += `${marked.parse(typeDefinition.doc)}\n`;
    }

    const annotations: [string, unknown][] = Object.entries(typeDefinition).filter(([key]) => key.startsWith("@"));
    const restProps = Object.entries(typeDefinition).filter(
      ([key]) => !key.startsWith("@") && !["doc", "kind", "type", "length"].includes(key),
    );

    if (!restProps.length && !annotations.length) return output;

    output += `<table>\n`;

    output += `<tr>`;
    output += `<th>Type</th>`;
    output += `<th>Description</th>`;
    output += `</tr>`;

    output += `<tr>`;
    const lengthConstraint =
      isLengthConstrainable(typeDefinition) && typeDefinition.length ? `(${typeDefinition.length})` : "";
    output += `<td>${typeDefinition.type}${lengthConstraint}</td>`;
    output += `<td>`;
    let description = "";
    const descriptionParts = restProps.map(
      ([key, value]) => `${key}: ${typeof value === "object" ? renderObject(value) : value}`,
    );
    const annotationParts = annotations.map(
      ([key, value]) =>
        `${key}: ${typeof value === "object" ? renderObject(value) : renderContentWithI18n(value, i18n)}`,
    );

    if (descriptionParts.length) {
      description += `${descriptionParts.join("<br />")}<br /><br />`;
    }
    if (annotationParts.length) {
      description += `<a href="${getExtensionsDocumentationLink()}" target="_blank">Annotations</a>:<br />`;
      description += `${annotationParts.join("<br />")}`;
    }
    output += description;
    output += `</td>`;
    output += `</tr>\n`;

    output += `</table>\n\n`;
  }
  return output;
}

function processServices(
  services: [string, ServiceDefinition][],
  entities: [string, EntityDefinition][],
  i18n: CSNInteropRoot["i18n"] | undefined,
): string {
  if (!services.length) return "";
  let output = "## Services\n\n";
  for (const [serviceName, serviceDefinition] of services) {
    output += `### ${serviceName}\n\n`;
    if (serviceDefinition.doc) {
      output += `${marked.parse(serviceDefinition.doc)}\n`;
    }

    const exposedEntities = entities.filter(([entityName]) => entityName.split(".")[0] === serviceName);
    if (exposedEntities) {
      output += `Exposed Entities:\n\n${exposedEntities.reduce((result, [entityName]) => [...result, `  - [${entityName}](#${getHeaderId(entityName)})`], [] as string[]).join("\n")}\n\n`;
    }

    const annotations: [string, unknown][] = Object.entries(serviceDefinition).filter(([key]) => key.startsWith("@"));
    const restProps = Object.entries(serviceDefinition).filter(
      ([key]) => !key.startsWith("@") && !["doc", "kind"].includes(key),
    );

    if (!restProps.length && !annotations.length) return output;

    output += `<table>\n`;

    output += `<tr>`;
    output += `<th>Description</th>`;
    output += `</tr>`;

    output += `<tr>`;
    output += `<td>`;
    let description = "";
    const descriptionParts = restProps.map(
      ([key, value]) => `${key}: ${typeof value === "object" ? renderObject(value) : value}`,
    );
    const annotationParts = annotations.map(
      ([key, value]) =>
        `${key}: ${typeof value === "object" ? renderObject(value) : renderContentWithI18n(value, i18n)}`,
    );

    if (descriptionParts.length) {
      description += `${descriptionParts.join("<br />")}<br /><br />`;
    }
    if (annotationParts.length) {
      description += `<a href="${getExtensionsDocumentationLink()}" target="_blank">Annotations</a>:<br />`;
      description += `${annotationParts.join("<br />")}`;
    }
    output += description;
    output += `</td>`;
    output += `</tr>\n`;

    output += `</table>\n\n`;
  }
  return output;
}

// Main renderer function
export const renderer = async (json: CSNInteropRoot, asHTml: boolean = false): Promise<string> => {
  const { definitions, meta, i18n } = json;

  let output = "";

  if (meta?.document?.title) output += `# ${meta.document.title}\n\n`;
  if (meta?.document?.doc) output += `${meta.document.doc}\n\n`;

  // Filter entities and types
  const entities = Object.entries(definitions).filter(([_, def]) => isEntityDefinition(def)) as [
    string,
    EntityDefinition,
  ][];

  const types = Object.entries(definitions).filter(([_, def]) => isTypeDefinition(def)) as [string, TypeDefinition][];

  const services = Object.entries(definitions).filter(([_, def]) => isServiceDefinition(def)) as [
    string,
    ServiceDefinition,
  ][];

  output += processEntities(
    entities,
    services.map((serviceTuple) => serviceTuple[0]),
    i18n,
  );
  output += processTypes(types, i18n);
  output += processServices(services, entities, i18n);

  if (asHTml) {
    marked.use(gfmHeadingId());
    const html = await marked.parse(output);
    return html;
  }

  return output;
};
