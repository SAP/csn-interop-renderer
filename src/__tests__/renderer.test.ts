import { renderer } from "../renderer.js";
import example from "../../examples/CSNInterop.js";

describe("Renderer Tests", () => {
  test("generates markdown output from example file", async () => {
    const result = await renderer(example);

    expect(result).toMatchSnapshot();
  });
});
