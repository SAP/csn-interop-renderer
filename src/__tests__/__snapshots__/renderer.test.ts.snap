// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer Tests generates markdown output from example file 1`] = `
"# Airline (Example)

This CSN example document shows how the airline example is expressed with a CDS **Service** exposing the entities through an API.

## Entity Definitions

### AirlineService.Airline

<p>Human readable description of the entity, in <strong>markdown</strong>.</p>

Exposed via:
[AirlineService](#airlineservice)

<table>
<tr><th>Element</th><th>Type</th><th>Description</th></tr><tr><td><strong id="airlineserviceairline-airlineid">AirlineID</strong><br /><br /><small>Human readable description of the element, in **markdown**.</small></td><td><a href="#airlineuuid">AirlineUuid</a></td><td>key: true<br /><br /><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Airline"<br />@ObjectModel.text.element: <code>["Name"]</code></td></tr>
<tr><td><strong id="airlineserviceairline-name">Name</strong><br /><br /><small></small></td><td>cds.String(4)</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Name"<br />@Semantics.text: true</td></tr>
<tr><td><strong id="airlineserviceairline-currencycode_code">CurrencyCode_code</strong><br /><br /><small></small></td><td>cds.String(17)</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Currency Code"<br />@Semantics.currencyCode: true</td></tr>
</table>

### AirlineService.Airport

Exposed via:
[AirlineService](#airlineservice)

<table>
<tr><th>Element</th><th>Type</th><th>Description</th></tr><tr><td><strong id="airlineserviceairport-airportid">AirportID</strong><br /><br /><small></small></td><td>cds.String(9)</td><td>key: true<br /><br /><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Airport"<br />@ObjectModel.text.element: <code>["Name"]</code></td></tr>
<tr><td><strong id="airlineserviceairport-name">Name</strong><br /><br /><small></small></td><td>cds.String(4)</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Name"<br />@Semantics.text: true</td></tr>
<tr><td><strong id="airlineserviceairport-city">City</strong><br /><br /><small></small></td><td>cds.String(4)</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "City"</td></tr>
<tr><td><strong id="airlineserviceairport-countrycode_code">CountryCode_code</strong><br /><br /><small></small></td><td>cds.String(16)</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Country Code"<br />@ObjectModel.foreignKey.association: <code>&lbrace;"=":"to_CountryCode"&rbrace;</code></td></tr>
<tr><td><strong id="airlineserviceairport-to_countrycode">to_CountryCode</strong><br /><br /><small></small></td><td>cds.Association</td><td>Association to one <a href="#airlineservicecountries">AirlineService.Countries</a> (path: <a href="#airlineservicecountries">AirlineService.Countries</a>.<a href="#airlineservicecountries-code">code</a>) via <a href="#airlineserviceairport-countrycode_code">CountryCode_code</a><br /><br /></td></tr>
</table>

### AirlineService.Countries

Exposed via:
[AirlineService](#airlineservice)

<table>
<tr><th>Element</th><th>Type</th><th>Description</th></tr><tr><td><strong id="airlineservicecountries-code">code</strong><br /><br /><small></small></td><td>cds.String(4)</td><td>key: true<br /><br /><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Country Code"<br />@ObjectModel.text.association: <code>&lbrace;"=":"texts"&rbrace;</code></td></tr>
<tr><td><strong id="airlineservicecountries-texts">texts</strong><br /><br /><small></small></td><td>cds.Composition</td><td></td></tr>
</table>

### AirlineService.Countries_texts

Exposed via:
[AirlineService](#airlineservice)

<table>
<tr><th>Element</th><th>Type</th><th>Description</th></tr><tr><td><strong id="airlineservicecountries_texts-code">code</strong><br /><br /><small></small></td><td>cds.String(4)</td><td>key: true<br /><br /><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Country Code"</td></tr>
<tr><td><strong id="airlineservicecountries_texts-locale">locale</strong><br /><br /><small></small></td><td>cds.String(6)</td><td>key: true<br /><br /><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Language Code"<br />@Semantics.language: true</td></tr>
<tr><td><strong id="airlineservicecountries_texts-name">name</strong><br /><br /><small></small></td><td>cds.String(4)</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Name"<br />@Semantics.text: true</td></tr>
<tr><td><strong id="airlineservicecountries_texts-descr">descr</strong><br /><br /><small></small></td><td>cds.String(5)</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Description"<br />@Semantics.text: true</td></tr>
</table>

### AirlineService.FlightConnection

Exposed via:
[AirlineService](#airlineservice)

<table>
<tr><th>Element</th><th>Type</th><th>Description</th></tr><tr><td><strong id="airlineserviceflightconnection-airlineid">AirlineID</strong><br /><br /><small></small></td><td>cds.String(9)</td><td>key: true<br /><br /><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Airline"<br />@ObjectModel.foreignKey.association: <code>&lbrace;"=":"to_Airline"&rbrace;</code></td></tr>
<tr><td><strong id="airlineserviceflightconnection-connectionid">ConnectionID</strong><br /><br /><small></small></td><td>cds.String(12)</td><td>key: true<br /><br /><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Flight Number"</td></tr>
<tr><td><strong id="airlineserviceflightconnection-departureairport_airportid">DepartureAirport_AirportID</strong><br /><br /><small></small></td><td>cds.String(26)</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Departure Airport"<br />@ObjectModel.foreignKey.association: <code>&lbrace;"=":"to_DepartureAirport"&rbrace;</code></td></tr>
<tr><td><strong id="airlineserviceflightconnection-destinationairport_airportid">DestinationAirport_AirportID</strong><br /><br /><small></small></td><td>cds.String(28)</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Destination Airport"<br />@ObjectModel.foreignKey.association: <code>&lbrace;"=":"to_DestinationAirport"&rbrace;</code></td></tr>
<tr><td><strong id="airlineserviceflightconnection-departuretime">DepartureTime</strong><br /><br /><small></small></td><td>cds.Time</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Departure Time"</td></tr>
<tr><td><strong id="airlineserviceflightconnection-arrivaltime">ArrivalTime</strong><br /><br /><small></small></td><td>cds.Time</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Arrival Time"</td></tr>
<tr><td><strong id="airlineserviceflightconnection-distance">Distance</strong><br /><br /><small></small></td><td>cds.Integer</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Distance"<br />@Semantics.quantity.unitOfMeasure: <code>&lbrace;"=":"DistanceUnit"&rbrace;</code></td></tr>
<tr><td><strong id="airlineserviceflightconnection-distanceunit">DistanceUnit</strong><br /><br /><small></small></td><td>cds.String(12)</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Distance Unit"</td></tr>
<tr><td><strong id="airlineserviceflightconnection-to_airline">to_Airline</strong><br /><br /><small></small></td><td>cds.Association</td><td>Association to many <a href="#airlineserviceairline">AirlineService.Airline</a> (path: <a href="#airlineserviceairline">AirlineService.Airline</a>.<a href="#airlineserviceairline-airlineid">AirlineID</a>) via <a href="#airlineserviceflightconnection-airlineid">AirlineID</a><br /><br /></td></tr>
<tr><td><strong id="airlineserviceflightconnection-to_departureairport">to_DepartureAirport</strong><br /><br /><small></small></td><td>cds.Association</td><td>Association to many <a href="#airlineserviceairport">AirlineService.Airport</a> (path: <a href="#airlineserviceairport">AirlineService.Airport</a>.<a href="#airlineserviceairport-airportid">AirportID</a>) via <a href="#airlineserviceflightconnection-departureairport_airportid">DepartureAirport_AirportID</a><br /><br /></td></tr>
<tr><td><strong id="airlineserviceflightconnection-to_destinationairport">to_DestinationAirport</strong><br /><br /><small></small></td><td>cds.Association</td><td>Association to many <a href="#airlineserviceairport">AirlineService.Airport</a> (path: <a href="#airlineserviceairport">AirlineService.Airport</a>.<a href="#airlineserviceairport-airportid">AirportID</a>) via <a href="#airlineserviceflightconnection-destinationairport_airportid">DestinationAirport_AirportID</a><br /><br /></td></tr>
</table>

### AirlineService.Flight

Exposed via:
[AirlineService](#airlineservice)

<table>
<tr><th>Element</th><th>Type</th><th>Description</th></tr><tr><td><strong id="airlineserviceflight-airlineid">AirlineID</strong><br /><br /><small></small></td><td>cds.String(9)</td><td>key: true<br /><br /><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Airline"<br />@ObjectModel.foreignKey.association: <code>&lbrace;"=":"to_Airline"&rbrace;</code></td></tr>
<tr><td><strong id="airlineserviceflight-flightdate">FlightDate</strong><br /><br /><small></small></td><td>cds.Date</td><td>key: true<br /><br /><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Flight Date"</td></tr>
<tr><td><strong id="airlineserviceflight-connectionid">ConnectionID</strong><br /><br /><small></small></td><td>cds.String(12)</td><td>key: true<br /><br /><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Flight Number"<br />@ObjectModel.foreignKey.association: <code>&lbrace;"=":"to_Connection"&rbrace;</code></td></tr>
<tr><td><strong id="airlineserviceflight-price">Price</strong><br /><br /><small></small></td><td>cds.Decimal</td><td>precision: 16<br />scale: 3<br /><br /><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Price"<br />@Aggregation.default: <code>&lbrace;"#":"MIN"&rbrace;</code><br />@Semantics.amount.currencyCode: <code>&lbrace;"=":"CurrencyCode_code"&rbrace;</code></td></tr>
<tr><td><strong id="airlineserviceflight-currencycode_code">CurrencyCode_code</strong><br /><br /><small></small></td><td>cds.String(17)</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Currency Code"<br />@Semantics.currencyCode: true</td></tr>
<tr><td><strong id="airlineserviceflight-planetype">PlaneType</strong><br /><br /><small></small></td><td>cds.String(9)</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Plane Type"</td></tr>
<tr><td><strong id="airlineserviceflight-maximumseats">MaximumSeats</strong><br /><br /><small></small></td><td>cds.Integer</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Maximum Seats"<br />@Aggregation.default: <code>&lbrace;"#":"SUM"&rbrace;</code></td></tr>
<tr><td><strong id="airlineserviceflight-occupiedseats">OccupiedSeats</strong><br /><br /><small></small></td><td>cds.Integer</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Occupied Seats"<br />@Aggregation.default: <code>&lbrace;"#":"SUM"&rbrace;</code></td></tr>
<tr><td><strong id="airlineserviceflight-to_airline">to_Airline</strong><br /><br /><small></small></td><td>cds.Association</td><td>Association to many <a href="#airlineserviceairline">AirlineService.Airline</a> (path: <a href="#airlineserviceairline">AirlineService.Airline</a>.<a href="#airlineserviceairline-airlineid">AirlineID</a>) via <a href="#airlineserviceflight-airlineid">AirlineID</a><br /><br /></td></tr>
<tr><td><strong id="airlineserviceflight-to_connection">to_Connection</strong><br /><br /><small></small></td><td>cds.Association</td><td>Association to many <a href="#airlineserviceflightconnection">AirlineService.FlightConnection</a> (path: <a href="#airlineserviceflightconnection">AirlineService.FlightConnection</a>.<a href="#airlineserviceflightconnection-airlineid">AirlineID</a>) via <a href="#airlineserviceflight-airlineid">AirlineID</a><br /><br /></td></tr>
</table>

### UnassignedEntity

<p>This entity is not prefixed with the service name and therefore not accessible through the service.</p>

<table>
<tr><th>Element</th><th>Type</th><th>Description</th></tr><tr><td><strong id="unassignedentity-someelement">SomeElement</strong><br /><br /><small></small></td><td>cds.String</td><td>key: true<br /><br /></td></tr>
</table>

## Type Definitions

### AirlineUuid

<p>Description of the reuse type</p>

<table>
<tr><th>Type</th><th>Description</th></tr><tr><td>cds.String(3)</td><td><a href="https://sap.github.io/csn-interop-specification/spec-v1/extensions/" target="_blank">Annotations</a>:<br />@EndUserText.label: "Airline"<br />@ObjectModel.text.element: <code>["Name"]</code></td></tr>
</table>

## Services

### AirlineService

<p>This is describing the service that exposes the CDS entities through an API.</p>

Exposed Entities:

  - [AirlineService.Airline](#airlineserviceairline)
  - [AirlineService.Airport](#airlineserviceairport)
  - [AirlineService.Countries](#airlineservicecountries)
  - [AirlineService.Countries_texts](#airlineservicecountries_texts)
  - [AirlineService.FlightConnection](#airlineserviceflightconnection)
  - [AirlineService.Flight](#airlineserviceflight)

"
`;
